File: python/bowling_game_test.py
import unittest
import bowling_game as bowling


class TestBowlingGame(unittest.TestCase):

    def setUp(self):
        self.game = bowling.BowlingGame()

    def testGutterGame(self):
        self._rollMany(20, 0)
        assert self.game.score() == 0

    def testAllOnes(self):
        self._rollMany(20, 1)
        assert self.game.score() == 20

    def testOneSpare(self):
        self.game.roll(5)
        self.game.roll(5)
        self.game.roll(3)
        self._rollMany(17, 0)
        assert self.game.score() == 16

    def testOneStrike(self):
        self.game.roll(10)
        self.game.roll(4)
        self.game.roll(3)
        self._rollMany(16, 0)
        assert self.game.score() == 24

    def testPerfectGame(self):
        self._rollMany(12, 10)
        assert self.game.score() == 300

    def testAllSpares(self):
        self._rollMany(21, 5)
        assert self.game.score() == 150
        
    def _rollMany(self, n, pins):
        for i in range(n):
            self.game.roll(pins)

File: cpp/bowling_game/bowling_game_test.cc
#include "gmock/gmock.h"
#include "bowling_game.h"

namespace {

class BowlingGameTest : public ::testing::Test {
 protected:
  BowlingGame g;

  void RollMany(int n, int pins) {
    for (int i = 0; i < n; i++) g.Roll(pins);
  }

  void RollSpare() {
    g.Roll(5);
    g.Roll(5);
  }

  void RollStrike() { g.Roll(10); }
};

TEST_F(BowlingGameTest, testGutterGame) {
  RollMany(20, 0);
  EXPECT_EQ(0, g.Score());
}

TEST_F(BowlingGameTest, testAllOnes) {
  RollMany(20, 1);
  EXPECT_EQ(20, g.Score());
}

TEST_F(BowlingGameTest, testOneSpare) {
  RollSpare();
  g.Roll(3);
  RollMany(17, 0);
  EXPECT_EQ(16, g.Score());
}

TEST_F(BowlingGameTest, testOneStrike) {
  RollStrike();
  g.Roll(3);
  g.Roll(4);
  RollMany(16, 0);
  EXPECT_EQ(24, g.Score());
}

TEST_F(BowlingGameTest, testPerfectGame) {
  RollMany(12, 10);
  EXPECT_EQ(300, g.Score());
}

}  // namespace
